### PR TITLE
HZN-1266: Add possibility to define src/dst port/address and exporter filter

### PR DESCRIPTION
**Pull request by mvrueden**
* JIRA: https://issues.opennms.org/browse/HZN-1266

The liquibase changelog files have changed, so running `install -dis` will fail. 
Here is how to update:

* Optional: If you have any custom rules defined, go to the Classification UI and export (download) the rules as CSV
* Stop OpenNMS
* Afterwards run the following SQL statements:
```
drop table classification_rules;
drop table classification_groups;
delete from databasechangelog where id in ('22.0.0-classification-tables', '22.0.0-classification-rules-and-groups-insert', '22.0.0-ruleNxtId');
```
* run `install -dis`
* start OpenNMS
* Optional: Migrate the downloaded rules (`2_rules.csv`) to the following schema: `name;protocol;srcAddress;srcPort;dstAddress;dstPort;exporterFilter`, where the mapping should be as follows:
  * `name` -> `name`
  * `protocol` -> `protocol`
  * `address` -> `dstAddress`
  * `port` -> `dstPort`
 Afterwards use the Upload function in the Classification UI to import the rules

Please squash-merge afterwards.

### CODE REVIEW

**Comment by mvrueden** [comment]
Yes, I am currently working on those.

**Comment by j-white** [comment]
Thanks, that helps a lot.

**Comment by mvrueden** [comment]
The tests should pass now.

Should I also add caching for the `exporterFilter` lookup? 
In general there will probably not be many of those, but caching them may be reasonable.

We also should consider removing the caching logic for the classification lookup, as with src and exporter address included in the request, a cache hit is very unlikely.

**Comment by j-white** [comment]
About the `exporterFilter ` look-ups: Yes, we will want to add some level of caching, since these queries can be expensive on large systems and we can accept that the results may not be exact for a few minutes. We can address this separately though.

I agree about caching the look-ups, these have very high cardinality, so the cache likely adds more overhead than anything.

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationEngine.java [hunk 1]
```diff
@@ -128,8 +178,33 @@ public String classify(ClassificationRequest classificationRequest) {
...
+        if (Objects.equals(srcPortClassifiers, dstPortClassifiers)) {
+            return dstPortClassifiers;
+        }
 
-    private List<Classifier> getClassifiers(ClassificationRequest request) {
-        return classifierPortList.get(request.getPort());
+        // If rules for src and dst port exist and they are not identical,
+        // they must be deduped (merging and sorting the classifiers)
+        // 1. Merge
+        final Set<Classifier> classifiers = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (mvrueden) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationEngine.java]: I misread the line you are refering to. 
Yes, in this case it totally makes sense to use a `SortedSet`.

[the remark this answers is not in the document]
Comment (mvrueden) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationEngine.java]: I just learned something new: When using TreeSet some classifiers where not added accordingly. This happened due to the fact that in order to use them correctly, the following contract must be fullfilled:
`a.compareTo(b) == 0 -> a.equals(b) == true`. Or in other words, an entry is only added to the set if `(!a.equals(b) && a.compareTo(b) == 0)`. See [Comparable Java Doc](https://docs.oracle.com/javase/6/docs/api/java/lang/Comparable.html). 

We only want to sort by priority so compareTo returns 0 but the objects are not equal, therefore we violate the contract, thus the object don't occur. When using a HashSet and ArrayList, we can navigate around this issue.

**Comment by mvrueden** [comment]
I removed the existing caching and introduced a filter caching as discussed. The tests should now hopefully pass. Bamboo was down yesterday, so I didn't see which tests were failing (-:

**Comment by j-white** [comment]
The table layout gets odd when using long filter expressions:
image

Aside from that, it looks pretty good.

I also played around with improving the performance of the `DefaultClassificationEngine#getClassifiers(ClassificationRequest request)` method, but couldn't come up with anything better then what's already there :)

**Pull request merged**